### PR TITLE
Support building without statistical analysis

### PR DESCRIPTION
**Pull request by harendra-kumar**
This PR is based on the previous, yet unmerged PR. It modularises the analysis and measurement code. I added a build flag too to compile without statistical analysis. When compiled without statistical analysis it defaults to quick mode.

The commits are better viewed individually rather than as a whole, they are pretty incremental in nature, each one is small (when logical changes are made) or simple (e.g. when code movement is done).

### CODE REVIEW

File: Gauge/Internal.hs [hunk 1]
```diff
@@ -71,8 +71,8 @@ runOne desc bm file = do
       Just prog -> gaugeIO $ runBenchmarkWith prog desc timeLimit quickMode
       Nothing -> gaugeIO $
         if quickMode
-        then runBenchmark bm threshold 2 0
-        else runBenchmark bm threshold 10 timeLimit
+        then runBenchmark bm 30 2 0
```
Comment (vincenthz) [inline Gauge/Internal.hs]: I don't think it's a good idea to remove the self-explanatory binding `threshold` in favor of hardcoding the value in 3 different places with no explanation. it would have been easier to move to internal as a constant relative to measurement instead of analysis.

**Comment by vincenthz** [comment]
I like that we can now build without any stats stuff and have a quick mode. definitely something I wanted to have. The PR was a bit hard to follow though

File: Gauge/Internal.hs (hunk 1 above)
Comment (harendra-kumar) [reply Gauge/Internal.hs]: I thought quite a bit when writing this and then finally discarded the binding. The reasoning was if I am using a binding for threshold then I should use a binding for number of samples as well (the other hardcoded argument). When I used that as well I thought I am adding too much for too little gain because the function definition explains the arguments and if we type them better later then it will become obvious.

Also, this is the only place where we would be using it. The other place i.e. `getOverhead` is an ugliness that should go away very soon.

**Comment by harendra-kumar** [comment]
Thanks for spending time on it and merging it quickly. Cheers to building a kickass perf tool! You have already done quite a bit of work by simplifying the code and making it amenable to changes and that's how I could make more changes quickly. But, there is a lot more we need to do. I have a couple of big rocks on mind, we'll discuss.
